KH-14939: Create a Dart wrapper for the React TestUtils library

jonasray-wf opened the pull request:
**Acceptance Criteria:**
- A class will be available to simulate all supported events documented at http://facebook.github.io/react/docs/events.html
- Methods will be available to call the methods defined at http://facebook.github.io/react/docs/test-utils.html
- Tests will be written to demonstrate that the wrapper works correctly

**Design notes:**
- This work has been started and is currently stored at https://github.com/matthewbalvanz-wf/react-dart/tree/test-utils
- As this is a wrapper we are implementing methods that will call into the React Javascript framework
- Because React is expecting a DOM to be present the tests would need to be run in a browser (Chromium) or when headless using content_shell
 
**Reviewers**: @matthewbalvanz-wf @tylersnavely-wf

jonasray-wf commented:
**Note**: I assume we will also want to update the documentation when we make the real PR

jonasray-wf commented on test/react_test_utils_test.dart:
```diff
@@ -0,0 +1,277 @@
+    component['_renderedChildren'] = renderedChildren;
+
+    return component;
+  };
```
I think `mockComponent()` is supposed to do the same thing. It is the last thing I ported over and I did some reading online to see how to use it. Most people said they ended up writing their own as well because `mockComponent()` depends on <a href='https://groups.google.com/forum/#!topic/reactjs/u6CbOF3lIig' target='_blank'>Jest</a>.

jonasray-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,227 @@
+/// Returns true if instance is a plain text component.
+bool isTextComponent(JsObject element) {
+  return _TestUtils.callMethod('isTextComponent', [element]);
+}
```
I currently have no tests for `isTextComponent()`. I couldn't figure out how it worked and it's not even listed on the <a href='http://facebook.github.io/react/docs/test-utils.html' target='_blank'>React Utils page</a>.

matthewbalvanz-wf replied:
I might have copied from their source instead of the documentation.  If its not listed on their page lets delete the method.

jonasray-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,227 @@
+/// Included in Dart for completeness
+List findAllInRenderedTree(JsObject inst, JsFunction test) {
+  return _TestUtils.callMethod('findAllInRenderedTree', [inst, test]);
+}
```
There aren't any functional tests to cover this right now. In order to make a functional test, we'd have to pass a JS function into the `test` arg. I was not able to find a way to do that.

~~However, all the `scryRendered...` functions call this~~

matthewbalvanz-wf replied:
They don't call this method, they call the JS method this method calls, but they perform that call outside of Dart.

jonasray-wf replied:
I just looked again, you are correct.

matthewbalvanz-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,227 @@
+/// number of matches besides one.
+JsObject findRenderedDOMComponentWithTag(JsObject element, String tag) {
+  return _TestUtils.callMethod(
+  'findRenderedDOMComponentWithTag', [element, tag]);
```
I think this needs more indentation?

matthewbalvanz-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,227 @@
+/// (created with React.createClass()) whose type is of a React componentClass.
+bool isCompositeComponentWithType(JsObject element, JsFunction componentClass) {
+  return _TestUtils.callMethod(
+  'isCompositeComponentWithType', [element, componentClass]);
```
I think this needs more indentation?

matthewbalvanz-wf commented on test/react_test_utils_test.dart:
```diff
@@ -0,0 +1,277 @@
+    proto['setState'] = () => null;
+
+    var props = new JsObject(_Object);
+    properties.forEach((k, v) => props[k] = v);
```
Is this what `JsObject.jsify(object)` does?  https://api.dartlang.org/apidocs/channels/be/dartdoc-viewer/dart:js.JsObject#id_JsObject-jsify

matthewbalvanz-wf commented on README.md:
```diff
@@ -153,3 +153,6 @@ class MyComponent extends Component {
 
+## Testing using React Test Utilities
+
+For more robust example take a look at [test/react_test_utils_test.dart](test).
```
For _a_ more robust example...

matthewbalvanz-wf replied:
I'd like to expand this more as a user looking at this README for the first time may not have any idea what TestUtils is or why they would want to use it.  We should have a quick summary of what they would use it for and links out to the React pages describing what TestUtils is and does.

jonasray-wf commented:
@matthewbalvanz-wf @tylersnavely-wf I added a working example to the readme

greglittlefield-wf commented:
This is funny. When I forked the main repo, I saw the repo network and was like "Hey, I wonder if anyone else already started TestUtils stuff.", but forgot to check. Then I checked, and found a `test-utils` branch, surprisingly with a fellow `-wf` username attached to it :).

I did a little bit of work on this stuff, some of it just basic wrapping, but then also tweaking how `ReactComponentFactory` works so that custom components can be type-checked properly (JS component constructors, not Dart types).

The work is pushed to https://github.com/greglittlefield-wf/react-dart/pull/1, and has a couple small tests written around it.

I think that `ReactComponentFactory` stuff would be valuable; thoughts on incorporating it?

jonasray-wf commented:
@greglittlefield-wf I think you may have solved my biggest headache with the `ReactComponentFactory` work. This is great. I'm going to try implementing your solution.

greglittlefield-wf commented:
Awesome! If you run into any issues, hit me up; I'd be more them happy to help.

trentgrover-wf commented:
this is awesome...that is all

jonasray-wf commented:
@greglittlefield-wf Updated the PR. We also added a proxy to native component creation.

greglittlefield-wf commented:
Awesome! Glad it could be applied to the built-in React components as well.

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+}
+
+/// Returns true if instance is a DOM component (such as a <div> or <span>).
+bool isDOMComponent(Element instance) {
```
This should be `(JsObject instance)`, since the argument will be a React instance, and not an actual DOM element. Changing it gets all the tests to pass! :smile:

jonasray-wf commented:
@greglittlefield-wf @tylersnavely-wf @matthewbalvanz-wf 

Ready for code review again.

tylersnavely-wf commented:
+1

matthewbalvanz-wf commented on README.md:
```diff
@@ -153,3 +153,56 @@ class MyComponent extends Component {
+
+void main() {
+  reactClient.setClientConfiguration();
+  var component = reactTestUtils.renderIntoDocument(myTestComponent({}));
```
I'd like to move this component initialization either into a setup method or the test itself.  While this works for a single test in an example it can lead to state pollution when writing multiple tests and I'd like the example to not lead people into a trap when they copy/paste it.

greglittlefield-wf replied:
This name is actually misleading, as it renders into a detached DOM node.

_From https://github.com/facebook/react/blob/v0.12.1/src/test/ReactTestUtils.js#L41:_

> ``` js
>  renderIntoDocument: function(instance) {
>    var div = document.createElement('div');
>    // None of our tests actually require attaching the container to the
>    // DOM, and doing so creates a mess that we rely on test isolation to
>    // clean up, so we're going to stop honoring the name of this method
>    // (and probably rename it eventually) if no problems arise.
>    // document.documentElement.appendChild(div);
>    return React.render(instance, div);
>  },
> ```

matthewbalvanz-wf replied:
It is, but its as you indicated the same name as the JS version.  I'm tempted to adopt their bad name to ensure that people reading articles about writing React tests in JS can directly port their knowledge over.

Would it make sense to keep it but add a `render` that is merely an alias?

jonasray-wf commented:
@matthewbalvanz-wf Readme has been updated.

matthewbalvanz-wf commented:
:+1:

jonasray-wf commented:
@greglittlefield-wf Does this look good to you?

greglittlefield-wf commented on lib/react_client.dart:
```diff
@@ -41,6 +41,15 @@ newJsMap(Map map) {
 typedef Component ComponentFactory();
 
+class ReactComponentFactoryProxy implements Function {
+  final Function _call;
```
This can be typed:

``` dart
final ReactComponentFactory _call;
```

greglittlefield-wf commented on lib/react_client.dart:
```diff
@@ -235,10 +244,7 @@ ReactComponentFactory _registerComponent(ComponentFactory componentFactory, [Ite
-   * return ReactComponentFactory which produce react component with set props and children[s]
-   */
-  return (Map props, [dynamic children]) {
+  var _call = (Map props, [dynamic children]) {
```
(Minor) This is inside a closure, so prefixing the name with an underscore to make it private doesn't do anything.

greglittlefield-wf commented on lib/react_client.dart:
```diff
@@ -275,17 +285,19 @@ ReactComponentFactory _registerComponent(ComponentFactory componentFactory, [Ite
+    return _React['createElement'].apply([name, newJsMap(props), children]);
   };
+
+  return new ReactComponentFactoryProxy(context['React']['DOM'][name], _call);
```
Also minor, but this can be `_React['DOM'][name]` instead of `context['React']['DOM'][name]`.

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+  for (String key in keys) {
+    if (!object.hasProperty(key)) {
+      throw 'Unable to resolve '
+          '$key in $base${keys.join('.')}}.\n$errorIfNotFound';
```
I think this was probably my bad, but this should be

``` dart
          '$key in $base.${keys.join('.')}}.\n$errorIfNotFound';
```

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+/// used as a primitive for other test utils
+///
+/// Included in Dart for completeness
+List findAllInRenderedTree(JsObject tree, ReactComponentFactory test) {
```
`test` shouldn't be typedef'd to `ReactComponentFactory`. Not sure what the right type should be. `JsFunction`?

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+}
+
+/// Returns true if element is any ReactElement.
+bool isElement(element) {
```
I think `element` can be typed to `JsObject` here.

jonasray-wf replied:
I think we removed the type on this one because it doesn't work on elements you get from the DOM.

greglittlefield-wf replied:
Right, so since should only work on React elements (which are instances of [the `ReactElement` JS constructor/class](https://github.com/matthewbalvanz-wf/react-dart/blob/76bbed05ed4bcb3acc3dd4203e49ac6aceb0aa66/lib/react_with_addons.js#L9867), and thus of the Dart type `JsObject`) , and not actual DOM nodes (`Element`/`DOMElement` Dart types), shouldn't it be typed to `JsObject`?

jonasray-wf replied:
Oh yeah, you are correct.

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+/// Render a Component into a detached DOM node in the document.
+JsObject renderIntoDocument(JsObject instance) {
+  var div = new DivElement();
+  return _TestUtils.callMethod('renderIntoDocument', [instance, div]);
```
There's actually no second argument to [this method](https://github.com/matthewbalvanz-wf/react-dart/blob/76bbed05ed4bcb3acc3dd4203e49ac6aceb0aa66/lib/react_with_addons.js#L14116), so the `div` variable can be eliminated.

greglittlefield-wf commented on test/react_test_utils_test.dart:
```diff
@@ -0,0 +1,213 @@
+    expect(results[2]['tagName'], equals('DIV'));
+  });
+
+  test('renderIntoDocucment', () {
```
Typo here.

greglittlefield-wf commented on test/react_test_utils_test.dart:
```diff
@@ -0,0 +1,213 @@
+      var h1Element = findRenderedDOMComponentWithTag(component, 'h1');
+
+      expect(isDOMComponent(h1Element), isTrue);
+    });
```
This test could be simplified:

``` dart
     test('returns true when argument is a DOM component', () {
      component = renderIntoDocument(h1({}));
      expect(isDOMComponent(component), isTrue);
     });
```

Doesn't make too much of a difference, though.

greglittlefield-wf commented on test/react_test_utils_test.dart:
```diff
@@ -0,0 +1,213 @@
+    });
+
+    test('returns false argument is not an element', () {
+      expect(isElement(sampleComponent), isFalse);
```
This `group` might read clearer if this line were:

``` dart
      expect(isElement(div), isFalse);
```

(Just a suggestion)

greglittlefield-wf commented on lib/react_test_utils.dart:
```diff
@@ -0,0 +1,255 @@
+class Simulate {
+
+  static blur(JsObject element, [Map eventData = const{}]) =>
+      _Simulate.callMethod('blur', [element, eventData]);
```
These `eventData` methods need to be translated to `JsObject` instances to work.

I started looking into this and fiddling around with the tests, and then got a little carried away... So, I've made a PR into this branch with some changes that should fix this: https://github.com/matthewbalvanz-wf/react-dart/pull/2

greglittlefield-wf commented:
@jonasray-wf Sorry it took so long for me to do an in-depth review; I somehow lost track of this CR.

Made some comments about some really minor stuff (which you don't _have_ to respond to), and then the one about the `eventData`/`JsObject` issue.

Other than those things, looks great; nice stuff!

jonasray-wf commented:
@greglittlefield-wf I have updated the PR. The only thing I didn't fix is the `isDOMComponent` test. We were trying to use our own components with elements in it for some examples.

greglittlefield-wf commented:
@jonasray-wf Cool, looks good. What about https://github.com/matthewbalvanz-wf/react-dart/pull/1#discussion-diff-27628754?

jonasray-wf commented:
@greglittlefield-wf @tylersnavely-wf @matthewbalvanz-wf 

I pulled in Greg's changes for testing the eventData methods and addressed all suggestions. Ready for final round of reviewing.

greglittlefield-wf commented:
Thanks for bearing with me on that feedback. Looks solid.

**+1**

tylersnavely-wf commented:
+1

matthewbalvanz-wf commented:
:+1:

jonasray-wf commented:
PR has been made back to cleandart https://github.com/cleandart/react-dart/pull/38
